### ISSUE TITLE
[Feature Request] Merge fragments while the download is ongoing for `-N`

### ISSUE THREAD

**Issue by Jules-A** [yt-dlp/yt-dlp#359]
## Checklist

- [ ] I'm reporting a broken site support issue
- [ ] I've verified that I'm running yt-dlp version **2021.05.20**
- [x] I've checked that all provided URLs are alive and playable in a browser
- [x] I've checked that all URLs and arguments with special characters are properly quoted or escaped
- [x] I've searched the bugtracker for similar bug reports including closed ones
- [x] I've read bugs section in FAQ

## Verbose log

```
[debug] Encodings: locale cp1252, fs utf-8, out utf-8, pref cp1252
[debug] yt-dlp version 2021.05.20 (exe)
[debug] Python version 3.9.5 (CPython 64bit) - Windows-10-10.0.19043-SP0
[debug] exe versions: ffmpeg N-102620-g3300625c6f, ffprobe N-102620-g3300625c6f

Latest stable versions of Windows 10, Python and dependencies with Yt-dlp built from master as of this post.

```

## Description

When using concurrent fragment option (-N), fragments are saved to disk and not added to the part file until every frag is downloaded despite the documentation saying this shouldn't happen. The Yt-dlp exe also increases memory proportional to the fragment size every time a new one is downloaded causing it to grow in to the GBs on large files.

Args: `https://www.crunchyroll.com/ah-my-buddha/episode-6-dont-transform-713589 -o "D:\Programs\YoutubeDL\crunchyroll.com\%(series)s - S%(season_number)sE%(episode_number)s - %(episode)s - %(epoch)s.%(ext)s" --embed-subs --sub-lang en --cookies cookies-crunchyroll-com.txt --user-agent "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:88.0) Gecko/20100101 Firefox/88.0" --fragment-retries "infinite" -N 3 --no-mtime --download-archive archive.txt
`

![Capture](https://user-images.githubusercontent.com/1760158/120077883-3dae3880-c0df-11eb-84ec-0fb13e846247.PNG)

**Comment by pukkandan** [comment]
> When using concurrent fragment option (-N), fragments are saved to disk and not added to the part file until every frag is downloaded

This is a known issue. the multithreaded downloader was fully implemented by @shirt-dev. I have never worked with multithreading, hence why I haven't really looked into improving it. But now that we are removing py2 support, I will look into a better implementation.

> despite the documentation saying this shouldn't happen

Where?

> The Yt-dlp exe also increases memory proportional to the fragment size every time a new one is downloaded causing it to grow in to the GBs on large files.

The multithread download will take more memory than normal download, but it shouldn't increase with the downloaded content. There must be some object that is left undeleted. I will investigate

**Comment by Jules-A** [comment]
>This is a known issue (aria2c also behaves like this #214). the multithreaded downloader was fully implemented by @shirt-dev. I have never worked with multithreading, hence why I haven't really looked into improving it. But now that we are removing py2 support, I will look into a better implementation.

Oh I see. Honestly it wouldn't be a big deal if the Windows file explorer wasn't a buggy pile of garbage... Unfortunately it is and every time a fragment is downloaded it un-selects the file you have selected, scrolls to the top, stops you editing file name or doing anything else with another file ect. every 500ms or so making it hard to do anything when 700 fragments are needed even for a tiny 350mb file.

> Where?

Ah, I misread --keep-fragments , thought it specifically said after the fragment is downloaded but there's no mention of it.

**Comment by pukkandan** [comment]
> Honestly it wouldn't be a big deal if the Windows file explorer wasn't a buggy pile of garbage... Unfortunately it is and every time a fragment is downloaded it un-selects the file you have selected, scrolls to the top, stops you editing file name or doing anything else with another file ect. every 500ms or so making it hard to do anything when 700 fragments are needed even for a tiny 350mb file.

This is the exact reason I implemented `--path temp:`

**Comment by Jules-A** [comment]
> This is the exact reason I implemented `--path temp:`. I was tired of windows explorer refreshing every time something downloaded

Yeah that would definitely help but for some reason I can't seem to get it working when trying to use a different path (D:\Temp\) . I think I'd need some example.

**Comment by pukkandan** [comment]
The memory issue should be fixed in https://github.com/yt-dlp/yt-dlp/commit/d89da64b1d21b7faf107d26104f3210ea01be7a1

**Comment by Jules-A** [comment]
> This is the exact reason I implemented `--path temp:`. I was tired of windows explorer refreshing every time something downloaded

Thanks, I didn't realise I could do that (it doesn't seem to be documented or I couldn't find it). It was tricky to set up though and in the end I had to do `--path temp:D:\Temp\`  It's a little confusing without quotes, I didn't think it would work without them on a absolute path so I didn't think to try it until a while later... 
I wasn't quite sure how to use separate output templates on it but I managed to do it in a way that it doesn't matter.

> The memory issue should be fixed in [d89da64](https://github.com/yt-dlp/yt-dlp/commit/d89da64b1d21b7faf107d26104f3210ea01be7a1)

Thanks. Ideally I'd like the fragments being written from memory to the part file so the disk isn't being written to twice. I guess that's the new goal of this issue now?

**Comment by fstirlitz** [comment]
> This is the exact reason I implemented `--path temp:`. I was tired of windows explorer refreshing every time something downloaded

Why not simply put fragment files in a subdirectory?

**Comment by Jules-A** [comment]
> I am not sure this is possible. Since we are multi-threading, the fragments download out of order, but appending can only be done sequentially. So they will either need to be kept in memory till all prior fragments are downloaded, or written to disk

Yeah but usually the fragments are pretty small so keeping them in memory until prior fragments are downloaded shouldn't really be an issue unless something like -N 100 is possible (in which case keeping in memory could simply be disabled)...

**Comment by Jules-A** [comment]
@pukkandan I know you changed this issue into feature request but I've noticed some other bugs with -N, related to my initial issue. 

Sometimes file will fail to download (or merge, I'm really not sure since I haven't seen any errors, though maybe I just missed them). This happened before I started using --path temp:D:\Temp\  but since using that, temporary files are getting left behind there for files that didn't completely download (in this case fragments 377-745 and respective .part and .ytdl files. I understand the default is to resume downloads with previously downloaded fragments but it should still clear up all the fragments for that respective download on completion. The skipping of download might not be a bug but I was curious as to what --fragment-retries "infinite" really does since I have it set, I assume it is regarding that fragment as unavailable and by default getting skipped but I want it to retry until it finds it instead (at least for a minute).

Command (temporary switched back to batch files since the frontend I was using didn't like some stuff):

`yt-dlp -a "CR todownload - valarian.txt" -o "crunchyroll.com\%%(series)s - S%%(season_number)sE%%(episode_number)s - %%(episode)s - %%(epoch)s.%%(ext)s" --embed-subs --sub-lang en --cookies cookies-crunchyroll-com.txt --user-agent "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:89.0) Gecko/20100101 Firefox/89.0" --fragment-retries "infinite" -N 3 --no-mtime --download-archive archive.txt --path temp:D:\Temp\`
